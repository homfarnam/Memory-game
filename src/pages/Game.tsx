import { useEffect, useState } from "react"
import moment from "moment"
import { useHistory } from "react-router"
import Card from "../components/Card/Card"
import { IProps } from "../interfaces/types"
import { shuffleImages } from "../utils/utils"

function Game() {
  const [images, setImages] = useState<IProps[]>([])

  const [selectedCard, setSelectedCard] = useState<number[]>([])

  const [matchedImages, setMatchedImages] = useState<number[]>([])

  const [score, setScore] = useState(180)

  const history = useHistory()

  useEffect(() => {
    fetch("https://tinyfac.es/api/data?limit=8")
      .then((res) => res.json())
      .then((data) => {
        setImages(shuffleImages([...data, ...data]))
      })
      .catch((err) => console.log(err))
  }, [])

  const matchImages = (i: number, j: number) => {
    if (images?.[i]?.url === images?.[j]?.url) {
      setMatchedImages([...matchedImages, i, j])
      setSelectedCard([])
      setScore((prev) => prev + 40)
    } else {
      setSelectedCard([...selectedCard, j])
      if (selectedCard.length === 1) {
        setScore((prev) => prev - 20)
      }
    }
  }

  const whichItem = (index: number) => {
    if (selectedCard.length === 2) {
      setSelectedCard([index])
    } else {
      matchImages(selectedCard[0], index)
    }
  }

  useEffect(() => {
    if (score <= 0) {
      alert("Game Over")

      setTimeout(() => {
        history.push("/score")
      }, 300)
    }
  }, [score, history])

  useEffect(() => {
    if (selectedCard.length === 2) {
      setTimeout(() => {
        setSelectedCard([])
      }, 3000)
    }
  }, [selectedCard])

  useEffect(() => {
    if (matchedImages.length === 16) {
      alert("You Win")
    }
  }, [matchedImages])

  useEffect(() => {
    const date = moment().format("MM/DD/YYYY")
    if (matchedImages.length === 16) {
      const user = localStorage.getItem("user")
      const users = JSON.parse(localStorage.getItem("users") || "[]")
      const id = Math.random() * 1000

      users.push({
        id: id,
        user: user,
        date: date,
        score: score,
      })
      localStorage.setItem("users", JSON.stringify(users))

      history.push("/score")
    }
  }, [matchedImages, history, score])

  return (
    <div className="flex flex-col min-h-screen w-full items-center justify-around text-white bg-gradient-to-br from-gray-600 via-teal-700 to-gray-800">
      <header className="w-full flex flex-col items-center justify-center space-y-3 my-5 mx-5 px-4 text-center md:px-0 md:mx-0">
        <h2 className="text-xl">Play the Flip card game</h2>
        <span>
          Select two cards with same content consequtively to make them vanish
        </span>
        <span>Score: {score}</span>
      </header>

      <div className="w-full h-auto flex justify-center">
        <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-4 col-span-4 gap-4">
          {images.map((item, i) => {
            return (
              <Card
                key={item.id + i}
                item={item}
                index={i}
                whichItem={whichItem}
                visible={selectedCard.includes(i) || matchedImages.includes(i)}
              />
            )
          })}
        </div>
      </div>
    </div>
  )
}

export default Game
