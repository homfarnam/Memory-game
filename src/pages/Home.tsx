import React, { useState } from "react"
import TextField from "@material-ui/core/TextField"
import { Button } from "@material-ui/core"
import { useEffect } from "react"
import { Formik, Form } from "formik"
import { Link } from "react-router-dom"

interface FormValues {
  username: string
}

const Home = () => {
  const initialValue = localStorage.getItem("user")
  const [user, setUser] = useState<string>(initialValue ? initialValue : "")
  const [isText, setIsText] = useState<boolean>(false)

  useEffect(() => {
    const value = localStorage.getItem("user")
    if (value) {
      setUser((prev) => value.toString())
      setIsText(true)
    }
  }, [])

  const initialValues: FormValues = {
    username: "",
  }

  return (
    <div className="w-full flex items-center justify-center">
      <div className="border p-5 flex flex-col items-center justify-between h-[400px]">
        <div>
          <h3>Click Play button to play the Memory Card! Game</h3>
          {isText ? (
            <span id="user" onClick={() => setIsText((prev) => !prev)}>
              Hello {user}
            </span>
          ) : (
            <Formik
              initialValues={initialValues}
              onSubmit={(values, actions) => {
                actions.setSubmitting(true)

                setUser((prevState) => {
                  localStorage.setItem("user", values.username)
                  return values.username
                })
                setIsText(true)
                actions.setSubmitting(false)
              }}
            >
              {({
                values,
                handleReset,
                handleSubmit,
                handleBlur,
                handleChange,
                errors,
              }) => (
                <Form
                  onReset={handleReset}
                  onSubmit={handleSubmit}
                  className="w-full h-full space-y-5 flex flex-col justify-around items-center"
                >
                  <div>
                    <TextField
                      id="username"
                      value={values.username}
                      label="User"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      type="text"
                      placeholder="Enter your name"
                    />
                  </div>
                </Form>
              )}
            </Formik>
          )}
        </div>

        <Button
          disabled={!isText}
          className="absolute bottom-0"
          variant="contained"
          color="primary"
          component={Link}
          to="/game"
          type="submit"
        >
          Play!
        </Button>
      </div>
    </div>
  )
}

export default Home
