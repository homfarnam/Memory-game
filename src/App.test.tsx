import React from 'react';
import { cleanup, render, screen, waitFor } from '@testing-library/react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from 'enzyme-to-json';
import Game from './pages/Game';
import Home from './pages/Home';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import userEvent from '@testing-library/user-event';

Enzyme.configure({ adapter: new Adapter() });
afterEach(cleanup);

describe('snapshot test', () => {
  it('snapshot test Home page', () => {
    const wrapper = Enzyme.shallow(<Home />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('snapshot test Game page', () => {
    const wrapper = Enzyme.shallow(<Game />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('test home page have a title', () => {
    const history = createMemoryHistory();

    const { getAllByText } = render(
      <Router history={history}>
        <Home />
      </Router>
    );

    const text = getAllByText('Click the Memory Card! Game');

    expect(text).toBeTruthy();
  });
});
