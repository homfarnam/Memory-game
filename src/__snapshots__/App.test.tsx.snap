// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test snapshot test Game page 1`] = `
<main
  className="flex flex-col min-h-screen w-full items-center justify-around text-white bg-gradient-to-br from-gray-600 via-teal-700 to-gray-800"
>
  <header
    className="w-full flex flex-col items-center justify-center space-y-3 my-5 mx-5 px-4 text-center md:px-0 md:mx-0"
  >
    <h2
      className="w-full text-2xl font-medium my-5 flex items-center justify-center"
    >
      Hi 
    </h2>
    <h3
      className="text-xl"
    >
      Play the Flip card game
    </h3>
    <span>
      Select two cards with same content consequtively to make them vanish
    </span>
    <span
      className="text-lg mt-2"
    >
      Score: 
      180
    </span>
  </header>
  <div
    className="w-full h-full flex justify-center"
  >
    <LoaderShow />
  </div>
</main>
`;

exports[`snapshot test snapshot test Home page 1`] = `
<div
  className="w-full flex items-center justify-center"
>
  <div
    className="border p-5 flex flex-col items-center justify-between h-[400px] my-10 "
  >
    <div>
      <h3
        className="flex items-center justify-center w-full text-xl font-medium"
      >
        Click Play button to play the Memory Card! Game
      </h3>
      <Formik
        initialValues={
          Object {
            "username": "",
          }
        }
        onSubmit={[Function]}
      >
        <Component />
      </Formik>
    </div>
    <WithStyles(ForwardRef(Button))
      className="absolute bottom-0"
      color="primary"
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      disabled={true}
      to="/game"
      type="submit"
      variant="contained"
    >
      Play!
    </WithStyles(ForwardRef(Button))>
  </div>
</div>
`;
